File: python/Prog_산모양타일링.py
import sys

sys.setrecursionlimit(1000000)

dp = [-1 for _ in range(200100)]
N = -1
tops = []
MOD = 10007 

def find_dp(n) :
    if n == N*2 or n == N*2+1 : return 1
    
    if dp[n] != -1 : return dp[n]
    
    if n % 2 == 0 :
        dp[n] = find_dp(n+1) + find_dp(n+2)
    else :
        dp[n] = find_dp(n+1) + find_dp(n+2)
        if tops[n//2] : dp[n] += find_dp(n+1)
        
    dp[n] %= MOD
    return dp[n]
        
        

def solution(n, tops_):
    global dp, N, tops
    N = n
    tops = tops_
    return find_dp(0)

File: java/src/Prog_산모양타일링.java
class Solution {
    public int MOD = 10007;
    
    public int solution(int n, int[] tops) {
        int[] dp = new int[200001];
        
        dp[0] = 1;
        if (tops[0] == 1) dp[1] = 3;
        else dp[1] = 2;
        
        for (int i=2;i<=2*n;i++) {
            if (i%2 == 1 && tops[i/2] == 1) dp[i] = (dp[i-1]*2 + dp[i-2]) % MOD;
            else dp[i] = (dp[i-1] + dp[i-2]) % MOD;
        }
        
        return dp[2*n];
    }
    
    
}